### PR TITLE
gRPC CLI Client

### PR DESCRIPTION
**Pull request by adalton**
This change does the following:
* Add a new client to abstract away communications with the gRPC server
* Refactors the test that was using the direct gRPC client to use the new client abstraction
* Replaces the quick-n-dirty CLI with a more robust version that also uses the client abstraction

### CODE REVIEW

**Comment by adalton** [comment]
# Testing
* Start the server
  ```
  $ sudo ./build/jobmanager
  ```

* List jobs. Expect none.
  ```
  $ ./jobctl list
  There are no jobs
  ```

* Start a short-lived job
  ```
  $ ./jobctl start -j date -c $(which date)
  +------+--------------------------------------+
  | NAME |                  ID                  |
  +------+--------------------------------------+
  | date | eea9ab73-b726-4b12-a1ff-5124bf4e53db |
  +------+--------------------------------------+
  ```

* Verify that the job completed successfully
  ```
  $ ./jobctl query eea9ab73-b726-4b12-a1ff-5124bf4e53db
  +------+--------------------------------------+---------+--------+-----------+--------+-------+
  | NAME |                  ID                  | RUNNING |  PID   | EXIT CODE | SIGNAL | ERROR |
  +------+--------------------------------------+---------+--------+-----------+--------+-------+
  | date | eea9ab73-b726-4b12-a1ff-5124bf4e53db | false   | 416111 |         0 |        |       |
  +------+--------------------------------------+---------+--------+-----------+--------+-------+
  ```

* Verify that the job shows up in the list
  ```
  $ ./jobctl list
  +------+--------------------------------------+---------+--------+-----------+--------+-------+
  | NAME |                  ID                  | RUNNING |  PID   | EXIT CODE | SIGNAL | ERROR |
  +------+--------------------------------------+---------+--------+-----------+--------+-------+
  | date | eea9ab73-b726-4b12-a1ff-5124bf4e53db | false   | 416111 |         0 |        |       |
  +------+--------------------------------------+---------+--------+-----------+--------+-------+
  ```

* View the output generated by the job to standard output
  ```
  $ ./jobctl stream eea9ab73-b726-4b12-a1ff-5124bf4e53db
  Mon Dec 20 22:27:06 EST 2021
  ```

* Try to start a job with the same name, expect failure
  ```
  $ ./jobctl start -j date -c $(which date)
  Error: rpc error: code = AlreadyExists desc = job exists
  Usage:
    jobctl start [flags]
  
  Examples:
  start -j myJob -c /usr/bin/find -- /dir -type f
  
  Flags:
    -c, --command string   The command for the job to run; must supply full path
    -h, --help             help for start
    -j, --jobName string   The name of the job to create; must be unique
  
  Global Flags:
        --hostPort string   The <hostName>:<portNumber> of the jobmanager server (default ":24482")
    -u, --userID string     The name of the user (selects the appropriate credential) (default "client1")
  ```

* Start a long-lived job
  ```
  $ ./jobctl start -j longrunning -c /bin/bash -- -c 'for ((i = 0; i < 60; ++i)); do echo $i; sleep 1; done'
  +-------------+--------------------------------------+
  |    NAME     |                  ID                  |
  +-------------+--------------------------------------+
  | longrunning | 91655432-b9ed-49d7-8b60-57ac8f5c7eff |
  +-------------+--------------------------------------+
  ```

* Verify the job is still running
  ```
  $ ./jobctl query 91655432-b9ed-49d7-8b60-57ac8f5c7eff
  +-------------+--------------------------------------+---------+--------+-----------+--------+-------+
  |    NAME     |                  ID                  | RUNNING |  PID   | EXIT CODE | SIGNAL | ERROR |
  +-------------+--------------------------------------+---------+--------+-----------+--------+-------+
  | longrunning | 91655432-b9ed-49d7-8b60-57ac8f5c7eff | true    | 416259 |           |        |       |
  +-------------+--------------------------------------+---------+--------+-----------+--------+-------+
  ```

* Verify that all jobs shows up in the list
  ```
  $ ./jobctl list
  +-------------+--------------------------------------+---------+--------+-----------+--------+-------+
  |    NAME     |                  ID                  | RUNNING |  PID   | EXIT CODE | SIGNAL | ERROR |
  +-------------+--------------------------------------+---------+--------+-----------+--------+-------+
  | date        | eea9ab73-b726-4b12-a1ff-5124bf4e53db | false   | 416111 |         0 |        |       |
  | longrunning | 91655432-b9ed-49d7-8b60-57ac8f5c7eff | true    | 416259 |           |        |       |
  +-------------+--------------------------------------+---------+--------+-----------+--------+-------+
  ```

* Watch the command generate output, and the stream terminate when the command
  finishes
  ```
  $ ./jobctl stream 91655432-b9ed-49d7-8b60-57ac8f5c7eff
  0
  1
  2
  3
  ...
  57
  58
  59
  ```

* Start another long-running job
  ```
  $ ./jobctl start -j tobekilled -c $(which sleep) 1h
  +------------+--------------------------------------+
  |    NAME    |                  ID                  |
  +------------+--------------------------------------+
  | tobekilled | 6c3986f7-9628-40c2-bec8-b9552c36e415 |
  +------------+--------------------------------------+
  ```

* Get the job's pid
  ```
  $ ./jobctl query 6c3986f7-9628-40c2-bec8-b9552c36e415
  +------------+--------------------------------------+---------+--------+-----------+--------+-------+
  | NAME       |                  ID                  | RUNNING |  PID   | EXIT CODE | SIGNAL | ERROR |
  +------------+--------------------------------------+---------+--------+-----------+--------+-------+
  | tobekilled | 6c3986f7-9628-40c2-bec8-b9552c36e415 | true    | 416955 |           |        |       |
  +------------+--------------------------------------+---------+--------+-----------+--------+-------+
  ```

* Stop the job
  ```
  $ ./jobctl stop 6c3986f7-9628-40c2-bec8-b9552c36e415
  ```

* Get the job's status.  Note that it now has a value in the `SIGNAL` column
  ```
  $ ./jobctl query 6c3986f7-9628-40c2-bec8-b9552c36e415
  +------------+--------------------------------------+---------+--------+-----------+--------+------------------+
  | NAME       |                  ID                  | RUNNING |  PID   | EXIT CODE | SIGNAL |      ERROR       |
  +------------+--------------------------------------+---------+--------+-----------+--------+------------------+
  | tobekilled | 6c3986f7-9628-40c2-bec8-b9552c36e415 | false   | 416955 |           | killed | [signal: killed] |
  +------------+--------------------------------------+---------+--------+-----------+--------+------------------+
  ```

* Start a job as a different non-admin user.  Here I'll use the same name as
  the first user -- that's OK.
  ```
  $ ./jobctl -u client2 start -j date -c $(which date)
  +------+--------------------------------------+
  | NAME |                  ID                  |
  +------+--------------------------------------+
  | date | b56b7ab5-bc70-47c3-9182-8c037f4a8892 |
  +------+--------------------------------------+
  ```

* Verify that `client1` cannot see `client2`'s job
  ```
  $ ./jobctl list
  +-------------+--------------------------------------+---------+--------+-----------+--------+------------------+
  |    NAME     |                  ID                  | RUNNING |  PID   | EXIT CODE | SIGNAL |      ERROR       |
  +-------------+--------------------------------------+---------+--------+-----------+--------+------------------+
  | date        | eea9ab73-b726-4b12-a1ff-5124bf4e53db | false   | 416111 |         0 |        |                  |
  | longrunning | 91655432-b9ed-49d7-8b60-57ac8f5c7eff | false   | 416259 |         0 |        |                  |
  | tobekilled  | 6c3986f7-9628-40c2-bec8-b9552c36e415 | false   | 416955 |           | killed | [signal: killed] |
  +-------------+--------------------------------------+---------+--------+-----------+--------+------------------+
  ```

* Verify that `administrator` can see all the jobs.  Note that the administrator's
  view includes the owner.
  ```
  $ ./jobctl -u administrator list
  +---------+-------------+--------------------------------------+---------+--------+-----------+--------+------------------+
  |  OWNER  |    NAME     |                  ID                  | RUNNING |  PID   | EXIT CODE | SIGNAL |      ERROR       |
  +---------+-------------+--------------------------------------+---------+--------+-----------+--------+------------------+
  | client1 | date        | eea9ab73-b726-4b12-a1ff-5124bf4e53db | false   | 416111 |         0 |        |                  |
  | client1 | longrunning | 91655432-b9ed-49d7-8b60-57ac8f5c7eff | false   | 416259 |         0 |        |                  |
  | client1 | tobekilled  | 6c3986f7-9628-40c2-bec8-b9552c36e415 | false   | 416955 |           | killed | [signal: killed] |
  | client2 | date        | b56b7ab5-bc70-47c3-9182-8c037f4a8892 | false   | 416667 |         0 |        |                  |
  +---------+-------------+--------------------------------------+---------+--------+-----------+--------+------------------+
  ```

* Start a long-running job as client2
  ```
  ./jobctl -u client2 start -j longrunning -c $(which sleep) 1h
  +-------------+--------------------------------------+
  |    NAME     |                  ID                  |
  +-------------+--------------------------------------+
  | longrunning | f81b22af-bb5d-4e71-b672-1536b09f7d23 |
  +-------------+--------------------------------------+
  ```

* Ensure `client1` cannot stop `client2`'s job
  ```
  $ ./jobctl stop f81b22af-bb5d-4e71-b672-1536b09f7d23
  Error: rpc error: code = NotFound desc = job not found
  Usage:
    jobctl stop [flags]

  Examples:
  jobctl stop 8de11b74-5cd9-4769-b40d-53de13faf77f

  Flags:
    -h, --help   help for stop

  Global Flags:
        --hostPort string   The <hostName>:<portNumber> of the jobmanager server (default ":24482")
    -u, --userID string     The name of the user (selects the appropriate credential) (default "client1")
  ```

* Ensure that `administrator` can stop `client2`'s job
  ```
  $ ./jobctl -u administrator stop f81b22af-bb5d-4e71-b672-1536b09f7d23
  $ ./jobctl -u client2 query f81b22af-bb5d-4e71-b672-1536b09f7d23
  +-------------+--------------------------------------+---------+--------+-----------+--------+------------------+
  |    NAME     |                  ID                  | RUNNING |  PID   | EXIT CODE | SIGNAL |      ERROR       |
  +-------------+--------------------------------------+---------+--------+-----------+--------+------------------+
  | longrunning | f81b22af-bb5d-4e71-b672-1536b09f7d23 | false   | 417450 |           | killed | [signal: killed] |
  +-------------+--------------------------------------+---------+--------+-----------+--------+------------------+
  ```

* Verify we can stream stdout
  ```
  ./jobctl start -j ip -c $(which ip)
  +------+--------------------------------------+
  | NAME |                  ID                  |
  +------+--------------------------------------+
  | ip   | 53c05283-3054-4cba-8665-813d4458fcee |
  +------+--------------------------------------+

  ./jobctl stream -s stderr 53c05283-3054-4cba-8665-813d4458fcee
  Usage: ip [ OPTIONS ] OBJECT { COMMAND | help }
         ip [ -force ] -batch filename
  where  OBJECT := { address | addrlabel | fou | help | ila | l2tp | link |
                     macsec | maddress | monitor | mptcp | mroute | mrule |
                     neighbor | neighbour | netconf | netns | nexthop | ntable |
                     ntbl | route | rule | sr | tap | tcpmetrics |
                     token | tunnel | tuntap | vrf | xfrm }
         OPTIONS := { -V[ersion] | -s[tatistics] | -d[etails] | -r[esolve] |
                      -h[uman-readable] | -iec | -j[son] | -p[retty] |
                      -f[amily] { inet | inet6 | mpls | bridge | link } |
                      -4 | -6 | -I | -D | -M | -B | -0 |
                      -l[oops] { maximum-addr-flush-attempts } | -br[ief] |
                      -o[neline] | -t[imestamp] | -ts[hort] | -b[atch] [filename] |
                      -rc[vbuf] [size] | -n[etns] name | -N[umeric] | -a[ll] |
                      -c[olor]}
  ```

File: pkg/command/jobctl/query.go [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+		return err
+	}
+	defer c.Close()
+	var lastError error
+
+	jobStatusList := make([]*jobmanager.JobStatus, 0, len(jobIDs))
+
+	for _, jobID := range jobIDs {
+		func() {
+			ctx, cancel := context.WithTimeout(cmd.Context(), shortOperationTimeout)
```
Comment (codingllama) [inline pkg/command/jobctl/query.go:55]: Suggestion: ditch the anonymous func and defer in the loop (seem alright here)? You could aways just cancel() after the query too.

Same for similar idioms.

File: server/jobmanager/serverv1/jobmanager_server.go [hunk 2]
```diff
@@ -176,12 +176,8 @@ func (s *jobmanagerServer) StreamOutput(
 
 	for {
 		select {
-		// We don't have a specific requirement to support a deadline for the
-		// stream operation from our client. But, since this is an API, someone
-		// might develop a client for it that does want to set a deadline.
-		// This will support that.
 		case <-response.Context().Done():
-			return context.DeadlineExceeded
+			return context.Canceled
```
Comment (codingllama) [inline server/jobmanager/serverv1/jobmanager_server.go:180]: This should do the right thing:

```suggestion
			return response.Context().Error()
```

**Comment by codingllama** [comment]
> # Testing
> * Start the server
> (...)

Thanks for the instructions. WDYT about them to the README?

**Review by codingllama — APPROVED** [review]

File: pkg/command/jobctl/query.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+		jobStatusList = append(jobStatusList, status)
+	}
+
+	renderJobStatusList(jobStatusList)
+
+	return lastError
+}
+
+func renderJobStatusList(jobStatus []*jobmanager.JobStatus) {
+	isAdmin := argUserID == "administrator"
```
Comment (rosstimothy) [inline pkg/command/jobctl/query.go:74]:
```suggestion
	isAdmin := argUserID == jobmanager.Superuser
```

File: pkg/client/jobmanager/jobmanager_client.go [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+}
+
+// StreamStdout invokes an RPC on the JobManager server to stream the standard
+// output of the job with the given jobID.  This function will block until either
+// (1) the context is interrupted, or (2) the job completes.
+func (c *Client) StreamStdout(ctx context.Context, jobID string, out io.Writer) error {
+	return c.stream(ctx, jobID, out, jobmanagerv1.OutputStream_OutputStream_STDOUT)
+}
+
+// StreamStdout invokes an RPC on the JobManager server to stream the standard
```
Comment (rosstimothy) [inline pkg/client/jobmanager/jobmanager_client.go:129]: nit: copy/paste error
```suggestion
// StreamStderr invokes an RPC on the JobManager server to stream the standard
```

File: pkg/client/jobmanager/jobmanager_client.go [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+	for {
+		output, err := grpcStream.Recv()
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return err
+		}
+
+		out.Write(output.Output)
```
Comment (rosstimothy) [inline pkg/client/jobmanager/jobmanager_client.go:183]: Should we handle the error here?

File: pkg/client/jobmanager/jobmanager_client.go [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+// error of the job with the given jobID.  This function will block until either
+// (1) the context is interrupted, or (2) the job completes.
+func (c *Client) StreamStderr(ctx context.Context, jobID string, out io.Writer) error {
+	return c.stream(ctx, jobID, out, jobmanagerv1.OutputStream_OutputStream_STDERR)
+}
+
+// Close closes the connection to the JobManager server.
+func (c *Client) Close() {
+	c.conn.Close()
+}
```
Comment (rosstimothy) [inline pkg/client/jobmanager/jobmanager_client.go:139]: Why not just make this an `io.Closer`?
```suggestion
func (c *Client) Close() error {
	return c.conn.Close()
}
```

**Review by rosstimothy — APPROVED** [review]

**Pull request merged**